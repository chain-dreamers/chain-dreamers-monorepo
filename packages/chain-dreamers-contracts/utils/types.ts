export type Path = {
  d: number;
  fill: number;
  stroke: number;
};

export type Trait = Path[];

export type Palettes = {
  d: string[];
  fill: string[];
  trait: Record<string, Path[]>;
  layerIndexes: number[];
};

export type PalettesStorage = {
  fillBytes: string;
  dBytes: string[];
  dBytesIndexes: string;
  dIndexesDebug: number[];
  traitBytes: string;
  traitBytesIndexes: string;
  layerIndexes: string;
};

export type Layer = {
  traitName: string;
  itemName: string;
  hexString: string;
  layerIndex: number;
  itemIndex: number;
};

export type Layers = Layer[];

export type Runner = {
  runnerId: number;
  dna: string;
};

export type Runners = Runner[];
