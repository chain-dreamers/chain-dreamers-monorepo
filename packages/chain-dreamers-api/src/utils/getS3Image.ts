import AWS from "aws-sdk";

const S3 = new AWS.S3();

export const getS3Image = async (
  folder: string,
  id: string
): Promise<string | null> => {
  try {
    const data = await S3.getObject({
      Bucket: process.env.API_ASSETS_BUCKET_NAME as string,
      Key: `${folder}/${id}.png`,
    }).promise();

    return data.Body?.toString("base64") || null;
  } catch (error) {
    return null;
  }
};
