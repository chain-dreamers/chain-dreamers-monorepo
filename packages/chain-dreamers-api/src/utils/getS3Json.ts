import AWS from "aws-sdk";

const S3 = new AWS.S3();

export const getS3Json = async (
  folder: string,
  id: string
): Promise<string | null> => {
  try {
    const data = await S3.getObject({
      Bucket: process.env.API_ASSETS_BUCKET_NAME as string,
      Key: `${folder}/${id}.json`,
    }).promise();

    return data.Body?.toString() || null;
  } catch (error) {
    return null;
  }
};
