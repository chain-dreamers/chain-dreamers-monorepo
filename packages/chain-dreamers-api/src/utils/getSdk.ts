import { ethers } from "ethers";
import {
  getMainSdk,
  getOriginalRunnerSdk as originalRunnerSdk,
} from "@dethcrypto/eth-sdk-client";

export const getOriginalRunnerSdk = () => {
  const rinkebyProvider = ethers.getDefaultProvider("mainnet", {
    etherscan: process.env.ETHERSCAN_API_KEY,
    infura: process.env.INFURA_ID,
    alchemy: process.env.ALCHEMY_API_KEY,
    pocket: process.env.POCKET_APPLICATION_KEY,
  });
  const defaultSigner = ethers.Wallet.createRandom().connect(rinkebyProvider);

  return originalRunnerSdk(defaultSigner);
};

export const getSdk = () => {
  const mainnetProvider = ethers.getDefaultProvider(process.env.NETWORK, {
    etherscan: process.env.ETHERSCAN_API_KEY,
    infura: process.env.INFURA_ID,
    alchemy: process.env.ALCHEMY_API_KEY,
    pocket: process.env.POCKET_APPLICATION_KEY,
  });
  const defaultSigner = ethers.Wallet.createRandom().connect(mainnetProvider);

  return getMainSdk(defaultSigner);
};
