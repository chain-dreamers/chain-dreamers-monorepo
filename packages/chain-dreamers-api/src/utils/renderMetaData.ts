import * as meta from "../../meta.json";

export type MetaData = {
  name: string;
  description: string;
  attributes: { trait_type: string; value: string }[];
  image: string;
};

export const renderMetaData = (id: string, tokenData: number[]): MetaData => {
  const metadata: MetaData = {
    name: `Dreamer #${id}`,
    description: `Runners run, but sometimes they dream. This is one of their dreams.`,
    attributes: [],
    image:
      process.env.API_DOMAIN_BASE_PATH === ""
        ? `https://${process.env.API_DOMAIN_NAME}/tokens/${id}/img`
        : `https://${process.env.API_DOMAIN_NAME}/${process.env.API_DOMAIN_BASE_PATH}/tokens/${id}/img`,
  };

  for (const index of tokenData) {
    if (meta[index]) {
      metadata.attributes.push(meta[index]);
    }
  }

  return metadata;
};
