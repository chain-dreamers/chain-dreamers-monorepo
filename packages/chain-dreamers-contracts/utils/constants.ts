import fs from "fs";
import { Layers, Palettes, PalettesStorage, Runners, SKU } from "./types";

export const MAX_CONTRACT_SIZE = 24_000;
export const PALETTES_FILE = "../../packages/chain-dreamers-image-processing/assets/palettes-hv.json";
export const PALETTES_ENCODED_FILE = "assets/palettes-encoded.json";
export const RUNNERS_DNA_FILE = "assets/runners-dna.json";
export const RUNNERS_LAYERS_FILE = "assets/runners-layers.json";
export const NONE_COLOR = "000001";
export const SKUS_FILE = "assets/candy-shop-skus.json";

export const loadPalettes = (): Palettes => {
  try {
    return JSON.parse(fs.readFileSync(PALETTES_FILE, "utf8"));
  } catch (e) {
    console.error(e);
    return {
      d: [],
      fill: [],
      trait: {},
      layerIndexes: [],
    };
  }
};

export const loadPalettesEncoded = (): PalettesStorage => {
  try {
    return JSON.parse(fs.readFileSync(PALETTES_ENCODED_FILE, "utf8"));
  } catch (e) {
    console.error(e);
    return {
      fillBytes: "",
      dBytes: [],
      dBytesIndexes: "",
      dIndexesDebug: [],
      traitBytes: "",
      traitBytesIndexes: "",
      layerIndexes: "",
    };
  }
};

export const loadRunnersDna = (): Runners => {
  try {
    return JSON.parse(fs.readFileSync(RUNNERS_DNA_FILE, "utf8"));
  } catch (e) {
    console.error(e);
    return [];
  }
};

export const loadRunnersLayers = (): Layers => {
  try {
    return JSON.parse(fs.readFileSync(RUNNERS_LAYERS_FILE, "utf8"));
  } catch (e) {
    console.error(e);
    return [];
  }
};

export const loadSkus = (): SKU[] => {
  try {
    return JSON.parse(fs.readFileSync(SKUS_FILE, "utf8"));
  } catch (e) {
    console.error(e);
    return [];
  }
};

export const chainRunnersBaseRendererAddress = (): string => {
  return process.env.CHAIN_RUNNERS_BASE_RENDERER_ADDRESS || "";
};
export const chainRunnersAddress = (): string => {
  return process.env.CHAIN_RUNNERS_ADDRESS || "";
};

// Deploy constants
export const TAGS = {
  INTEGERS: "Integers",
  CHAIN_DREAMERS: "ChainDreamers",
  DREAMERS_RENDERER: "DreamersRenderer",
  DREAMERS_PALETTES: "ChainDreamersPalettes",
  CHAIN_RUNNERS: "ChainRunners",
  CHAIN_RUNNERS_TRAITS: "ChainRunnersTraits",
  BYTES_SHIFT: "BytesShift",
  CANDY_SHOP: "CandyShop",
  CANDY_SHOP_INVENTORY: "CandyShopInventory",
};
