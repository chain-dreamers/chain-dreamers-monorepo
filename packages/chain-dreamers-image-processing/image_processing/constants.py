from pathlib import Path

ASSETS_DIR = Path("assets")
VALISE_DIR = ASSETS_DIR / "VALISE_CHAIN_DREAMERS_TRAITS"
VALISE_COMPUTED_DIR = ASSETS_DIR / "VALISE_CHAIN_DREAMERS_COMPUTED"
PALETTES_FILE = ASSETS_DIR / "palettes.json"
PALETTES_HV_FILE = ASSETS_DIR / "palettes-hv.json"
NONE_COLOR = "#000001"
# See ChainRunnersBaseRendering.sol contract
TRAITS_ORDERING = [
    "BACKGROUND",
    "RACE",
    "FACE",
    "MOUTH",
    "NOSE",
    "EYES",
    "EAR_ACCESSORIES",
    "FACE_ACCESSORIES",
    "MASK",
    "HEAD_BELOW",
    "EYES_ACCESSORIES",
    "HEAD_ABOVE",
    "MOUTH_ACCESSORIES",
]
