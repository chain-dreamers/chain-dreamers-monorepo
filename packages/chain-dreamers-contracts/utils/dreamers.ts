import { Path, Trait } from "./types";
import { loadPalettes } from "./constants";

const palettes = loadPalettes();
export const D_ATTRIBUTE_PALETTE = ["M", "L", "Q", "C", "A", "H", "V", "Z"];
export const D_REGEX = /[MLQCAHVZ][\d ,.]+/g;

export const getSvg = (trait: Trait): string => {
  return trait
    .map(
      ({ d, fill }) =>
        `<path d='${palettes.d[d]}' fill='#${palettes.fill[fill]}' stroke='#000' stroke-linecap='round' stroke-linejoin='round' stroke-width='0.71' />`
    )
    .join("\n");
};

export const normalizeD = (d: string): string => {
  return d
    .replace(/%20/g, " ")
    .replace(D_REGEX, (path) => {
      const command = path[0];
      const args = path.slice(1).split(/[ ,]/g);
      return `${command} ${args
        .filter(Boolean)
        .map((n) => parseInt(n))
        .join(" ")} `;
    })
    .trim();
};

export const inlineTrait = (trait: Trait): string => {
  return trait
    .map(({ d, fill, stroke }) => `d: ${d}; fill: ${fill}; stroke: ${stroke};`)
    .join(" ")
    .trim();
};

export const encodeDAttribute = (attribute: string): string => {
  return ("000" + D_ATTRIBUTE_PALETTE.indexOf(attribute).toString(2)).slice(-3);
};

export const encodeDCoordinate = (coordinate: string): string => {
  if (parseInt(coordinate) > 255) {
    throw new Error("Coordinate is too large");
  }
  return coordinate
    ? ("0".repeat(8) + parseInt(coordinate).toString(2)).slice(-8)
    : "";
};

export const encodeDToBits = (path: string): string => {
  const pathElements = path.match(D_REGEX);
  if (pathElements === null || pathElements.join("") !== path) {
    throw new Error(`Invalid path string: ${path}`);
  }
  return pathElements
    .map(
      (item) =>
        encodeDAttribute(item[0]) +
        item.slice(1).split(/[ ,]/).map(encodeDCoordinate).join("")
    )
    .join("");
};

export const encodeDToBytes = (string: string): string => {
  const bits = encodeDToBits(string);
  const bytes = (bits + "0".repeat(8))
    .slice(0, 8 * Math.ceil(bits.length / 8))
    .match(/.{8}/g);

  if (bytes === null || bytes.join("").slice(0, bits.length) !== bits) {
    throw new Error("Invalid path string");
  }

  return bytes
    .map((byte) => ("0" + parseInt(byte, 2).toString(16)).slice(-2))
    .join("");
};

export const uint16ToBytes = (uint16: number): string => {
  return ("0".repeat(4) + uint16.toString(16)).slice(-4);
};

export const numberToUint =
  (bits: number) =>
  (number: number): string => {
    return ("0".repeat(bits) + number.toString(2)).slice(-bits);
  };

export const encodePath = (path: Path): string => {
  return (
    "0".repeat(6) +
    parseInt(
      numberToUint(12)(path.d) + numberToUint(11)(path.fill) + path.stroke,
      2
    ).toString(16)
  ).slice(-6);
};

export const encodeTrait = (trait: Trait): string => {
  return trait.map(encodePath).join("");
};
