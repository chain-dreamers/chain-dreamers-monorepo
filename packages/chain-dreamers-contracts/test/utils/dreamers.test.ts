import chai, { expect } from "chai";
import {
  encodeDAttribute,
  encodeDCoordinate,
  encodeDToBits,
  encodeDToBytes,
  encodePath,
  encodeTrait,
  getSvg,
  normalizeD,
  numberToUint,
  uint16ToBytes,
} from "../../utils/dreamers";
import { Path, Trait } from "../../utils/types";
import { jestSnapshotPlugin } from "mocha-chai-jest-snapshot";
import { loadPalettes } from "../../utils/constants";

const palettes = loadPalettes();
chai.use(jestSnapshotPlugin());

describe("dreamers utils", function () {
  describe("normalizeD", () => {
    const tests = [
      {
        args: "M 0.0,0.0 L 255.0,0.0 L 255.0,255.0 L 0.0,255.0 L 0.0,0.0",
        expected: "M 0 0 L 255 0 L 255 255 L 0 255 L 0 0",
      },
      {
        args: "M 0.0,118.0 C 0.0,118.0 32.0,274.0 37.0,178.0 C 42.0,82.0 81.0,0.0 78.0,77.0 C 75.0,195.0 102.0,276.0 112.0,194.0 C 123.0,112.0 156.0,109.0 156.0,182.0 L 156.0,255.0 L 0.0,255.0 L 0.0,118.0",
        expected:
          "M 0 118 C 0 118 32 274 37 178 C 42 82 81 0 78 77 C 75 195 102 276 112 194 C 123 112 156 109 156 182 L 156 255 L 0 255 L 0 118",
      },
    ];
    tests.forEach((test) => {
      it(`should normalize ${test.args}`, () => {
        expect(normalizeD(test.args)).to.equal(test.expected);
      });
    });
  });
  describe("encodeDAttribute", () => {
    const tests = [
      { args: "M", expected: "000" },
      { args: "L", expected: "001" },
      { args: "Q", expected: "010" },
      { args: "C", expected: "011" },
      { args: "A", expected: "100" },
      { args: "H", expected: "101" },
      { args: "V", expected: "110" },
      { args: "Z", expected: "111" },
    ];

    tests.forEach((test) => {
      it(`should encode ${test.args} as ${test.expected}`, () => {
        expect(encodeDAttribute(test.args)).to.equal(test.expected);
      });
    });
  });
  describe("encodeDCoordinate", () => {
    const tests = [
      { args: "0", expected: "00000000" },
      { args: "1", expected: "00000001" },
      { args: "2", expected: "00000010" },
      { args: "255", expected: "11111111" },
    ];

    tests.forEach((test) => {
      it(`should encode ${test.args} as ${test.expected}`, () => {
        expect(encodeDCoordinate(test.args)).to.equal(test.expected);
      });
    });
  });
  describe("encodeDToBits", () => {
    const tests = [
      { args: "M 0 255", expected: "000" + "00000000" + "11111111" },
      { args: "L 0 255", expected: "001" + "00000000" + "11111111" },
      { args: "Q 0 255", expected: "010" + "00000000" + "11111111" },
      { args: "C 0 255", expected: "011" + "00000000" + "11111111" },
      {
        args: "M 0 255L10 20",
        expected:
          "000" + "00000000" + "11111111" + "001" + "00001010" + "00010100",
      },
      {
        args: "M 0.0,0.0 L 255.0,0.0 L 255.0,255.0 L 0.0,255.0 L 0.0,0.0",
        expected:
          "000" +
          "00000000" +
          "00000000" +
          "001" +
          "11111111" +
          "00000000" +
          "001" +
          "11111111" +
          "11111111" +
          "001" +
          "00000000" +
          "11111111" +
          "001" +
          "00000000" +
          "00000000",
      },
    ];

    tests.forEach((test) => {
      it(`should encode ${test.args} as ${test.expected}`, () => {
        expect(encodeDToBits(test.args)).to.equal(test.expected);
      });
    });
  });
  describe("encodeDToBytes", () => {
    const tests = [
      {
        args: "M 0 255",
        expected:
          "00" +
          parseInt("000" + "00000000" + "11111111" + "00000", 2).toString(16),
      },
      {
        args: "L 0 255",
        expected: parseInt(
          "001" + "00000000" + "11111111" + "00000",
          2
        ).toString(16),
      },
      {
        args: "Q 0 255",
        expected: parseInt(
          "010" + "00000000" + "11111111" + "00000",
          2
        ).toString(16),
      },
      {
        args: "C 0 255",
        expected: parseInt(
          "011" + "00000000" + "11111111" + "00000",
          2
        ).toString(16),
      },
      {
        args: "M 0 255L10 20",
        expected:
          "00" +
          parseInt(
            "000" +
              "00000000" +
              "11111111" +
              "001" +
              "00001010" +
              "00010100" +
              "00",
            2
          ).toString(16),
      },
    ];

    tests.forEach((test) => {
      it(`should encode ${test.args} as ${test.expected}`, () => {
        expect(encodeDToBytes(test.args)).to.equal(test.expected);
      });
    });
  });
  describe("uint16ToBytes", () => {
    const tests = [
      { args: 0, expected: "0000" },
      { args: 10, expected: "000a" },
      { args: 100, expected: "0064" },
      { args: 1_000, expected: "03e8" },
      { args: 10_000, expected: "2710" },
      { args: 65_535, expected: "ffff" },
    ];

    tests.forEach((test) => {
      it(`should encode ${test.args} as ${test.expected}`, () => {
        expect(uint16ToBytes(test.args)).to.equal(test.expected);
      });
    });
  });
  describe("numberToUint", () => {
    const tests = [
      { args: 0, expected: "000000000000" },
      { args: 1, expected: "000000000001" },
      { args: 2, expected: "000000000010" },
      { args: 255, expected: "000011111111" },
      { args: 4095, expected: "111111111111" },
    ];

    tests.forEach((test) => {
      it(`should encode ${test.args} as ${test.expected}`, () => {
        expect(numberToUint(12)(test.args)).to.equal(test.expected);
      });
    });
  });
  describe("encodePath", () => {
    const tests = [
      { args: { d: 0, fill: 0 } as Path, expected: "000000" },
      { args: { d: 15, fill: 0 } as Path, expected: "00f000" },
      { args: { d: 15, fill: 15 } as Path, expected: "00f00f" },
      { args: { d: 17, fill: 200 } as Path, expected: "0110c8" },
      { args: { d: 4095, fill: 4095 } as Path, expected: "ffffff" },
    ];
    tests.forEach((test) => {
      it(`should encode ${test.args} as ${test.expected}`, () => {
        expect(encodePath(test.args)).to.equal(test.expected);
      });
    });
  });
  describe("encodeTrait", () => {
    const tests = [
      {
        args: [
          { d: 0, fill: 0 },
          { d: 15, fill: 0 },
        ] as Trait,
        expected: "00000000f000",
      },
      {
        args: [
          { d: 15, fill: 15 },
          { d: 17, fill: 200 },
          { d: 4095, fill: 4095 },
        ] as Trait,
        expected: "00f00f0110c8ffffff",
      },
    ];
    tests.forEach((test) => {
      it(`should encode ${test.args} as ${test.expected}`, () => {
        expect(encodeTrait(test.args)).to.equal(test.expected);
      });
    });
  });
  describe("getSvg", () => {
    it("Should match snapshot", () => {
      expect(getSvg(palettes.trait[0])).toMatchSnapshot();
    });
  });
});
