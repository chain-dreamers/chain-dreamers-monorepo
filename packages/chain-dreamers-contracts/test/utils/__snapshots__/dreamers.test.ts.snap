// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dreamers utils getSvg Should match snapshot 1`] = `
"<path d='M 0.0,0.0 H 255.0 V 255.0 H 0.0 V 0.0' fill='#ffffff' stroke='#000' stroke-linecap='round' stroke-linejoin='round' stroke-width='0.71' />
<path d='M 41.0,255.0 H 102.0 C 102.0,255.0 137.0,204.0 102.0,138.0 C 67.0,72.0 132.0,0.0 132.0,0.0 H 41.0 V 255.0' fill='#7acbe1' stroke='#000' stroke-linecap='round' stroke-linejoin='round' stroke-width='0.71' />
<path d='M 0.0,255.0 H 61.0 C 61.0,255.0 96.0,204.0 61.0,138.0 C 26.0,72.0 91.0,0.0 91.0,0.0 H 0.0 V 255.0' fill='#6d9cf8' stroke='#000' stroke-linecap='round' stroke-linejoin='round' stroke-width='0.71' />
<path d='M 214.0,0.0 H 153.0 C 153.0,0.0 118.0,51.0 153.0,117.0 C 188.0,182.0 123.0,255.0 123.0,255.0 H 214.0 V 0.0' fill='#7acbe1' stroke='#000' stroke-linecap='round' stroke-linejoin='round' stroke-width='0.71' />
<path d='M 255.0,0.0 H 194.0 C 194.0,0.0 159.0,51.0 194.0,117.0 C 229.0,182.0 164.0,255.0 164.0,255.0 H 255.0 V 0.0' fill='#6d9cf8' stroke='#000' stroke-linecap='round' stroke-linejoin='round' stroke-width='0.71' />"
`;
